### PR TITLE
Print module

### PR DESCRIPTION
**Pull request by TurboGit**
This is latest version which is probably ready for merging. The code is ported to Gtk3 and using the PDF backend (thanks to Tobias!). After good report from the mailing-list I'm confident that it is ready.

### CODE REVIEW

File: src/common/cups_print.c [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+
+    // check that the printer is ready
+    if (strtol(psvalue, NULL, 10) < IPP_PRINTER_STOPPED)
+    {
+      dt_printer_info_t *pr = dt_get_printer_info(dest->name);
+      result = g_list_append(result,pr);
+    }
+    else if (darktable.unmuted & DT_DEBUG_PRINT)
+    {
+      printf("[print] skip printer %s as stopped", dest->name);
```
Comment (levitte) [inline src/common/cups_print.c]: nit pick: end the string with \n

Comment (TurboGit) [inline src/common/cups_print.c]: Fixed!

**Comment by TurboGit** [comment]
I can also report that the color-management for printer profile is working. Tested today.

File: src/common/cups_print.c [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with darktable.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <glib.h>
+#include <stdio.h>
+#include <tiffio.h>
```
Comment (LebedevRI) [inline src/common/cups_print.c]: Is this include needed? That is the only line with tiff substring in this file.

Comment (TurboGit) [inline src/common/cups_print.c]: Not needed anymore indeed! Good catch. Removed locally.

File: src/common/printprof.c [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with darktable.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <glib.h>
+#include <unistd.h>
+#include "lcms2_plugin.h"
```
Comment (LebedevRI) [inline src/common/printprof.c]: #include <lcms2_plugin.h> ?

File: src/views/view.c [hunk 4]
```diff
@@ -70,7 +70,11 @@ void dt_view_manager_init(dt_view_manager_t *vm)
 #ifdef HAVE_MAP
                       "map",
 #endif
-                      "slideshow",  NULL };
+                      "slideshow",
```
Comment (LebedevRI) [inline src/views/view.c]: Unrelated, but probably slideshow should be guarded by ifdefs too

File: src/views/view.c [hunk 5]
```diff
@@ -1222,6 +1226,85 @@ int dt_view_image_expose(dt_view_image_over_t *image_over, uint32_t imgid, cairo
   return missing;
 }
 
+void
```
Comment (LebedevRI) [inline src/views/view.c]: I do not really like this.
If this is _really_ needed, *i* would prefer if you would add/pass another param to dt_view_image_expose() that would force-disable display of meta-data.

Comment (TurboGit) [inline src/views/view.c]: I did that at first and it makes the dt_view_image_expose() even more unreadable. So to me it's a compromise between readability and code duplication. Would be good to have other point of view as I've no strong opinion and both solutions have pro and cons.

Comment (TurboGit) [inline src/views/view.c]: Yet, I'll see tomorrow if I can put the common code in a specific procedure. I don't remember all the details as this was done long time ago now.

Comment (TurboGit) [inline src/views/view.c]: @LebedevRI, I have refactored the code, what do you think? I find that the routine is a bit complex now but no strong opinion.

File: src/common/cups_print.c [hunk 6]
```diff
@@ -0,0 +1,538 @@
...
+
+      ppd_attr_t *attr = ppdFindAttr(ppd, "HWMargins", NULL);
+
+      if (attr)
+      {
+        sscanf(attr->value, "%lf %lf %lf %lf",
+               &result->hw_margin_left, &result->hw_margin_bottom,
+               &result->hw_margin_right, &result->hw_margin_top);
+
+        result->hw_margin_left   = dt_pdf_point_to_mm (result->hw_margin_left);
```
Comment (houz) [inline src/common/cups_print.c]: I don't think that you should be using dt_pdf_point_to_mm here, or is CUPS reporting the margings in PDF points (1/72 inch)?

File: src/views/view.c (hunk 5 above)
Comment (LebedevRI) [inline src/views/view.c]: Well, at lest this way that code is not duplicated. This looks better, at least to me.

File: src/common/cups_print.c [hunk 7]
```diff
@@ -0,0 +1,538 @@
...
+  int num_dests = cupsGetDests(&dests);
+  cups_dest_t *dest = cupsGetDest(printer_name, NULL, num_dests, dests);
+  dt_printer_info_t *result = NULL;
+
+  if (dest)
+  {
+    const char *PPDFile = cupsGetPPD (printer_name);
+    result = (dt_printer_info_t *)malloc(sizeof(dt_printer_info_t));
+    g_strlcpy(result->name, dest->name, MAX_NAME);
+    ppd_file_t *ppd = ppdOpenFile(PPDFile);
```
Comment (houz) [inline src/common/cups_print.c]: According to the docs the PPD API is deprecated and will be removed in the future: https://www.cups.org/documentation.php/doc-2.0/api-ppd.html

File: src/common/cups_print.c [hunk 8]
```diff
@@ -0,0 +1,538 @@
...
+          sscanf (attr->value, "%ddpi", &result->resolution);
+      }
+      else
+        result->resolution = 300;
+
+      while(result->resolution>360)
+        result->resolution /= 2.0;
+
+      ppdClose(ppd);
+      unlink(PPDFile);
```
Comment (houz) [inline src/common/cups_print.c]: Do you really have to unlink the PPD file? Do you get a custom copy passed instead of the system file?

File: src/common/cups_print.c [hunk 9]
```diff
@@ -0,0 +1,538 @@
...
+
+    // check that the printer is ready
+    if (strtol(psvalue, NULL, 10) < IPP_PRINTER_STOPPED)
+    {
+      dt_printer_info_t *pr = dt_get_printer_info(dest->name);
+      result = g_list_append(result,pr);
+    }
+    else if (darktable.unmuted & DT_DEBUG_PRINT)
+    {
+      printf("[print] skip printer %s as stopped\n", dest->name);
```
Comment (houz) [inline src/common/cups_print.c]: all these error messages should be fprintf(stderr, ...) instead.

File: src/common/cups_print.c [hunk 10]
```diff
@@ -0,0 +1,538 @@
...
+      {
+        if (cupsGetDestMediaByIndex(hcon, dest, info, k, CUPS_MEDIA_FLAGS_DEFAULT, &size))
+        {
+          if (!paper_exists(result,size.media))
+          {
+            pwg_media_t *med = pwgMediaForPWG (size.media);
+            char common_name[MAX_NAME] = { 0 };
+
+            if (med->ppd)
+              g_strlcpy(common_name, med->ppd, MAX_NAME);
```
Comment (houz) [inline src/common/cups_print.c]: Better use sizeof().

File: src/common/cups_print.c [hunk 11]
```diff
@@ -0,0 +1,538 @@
...
+  const char *PPDFile = cupsGetPPD(printer_name);
+  ppd_file_t *ppd = ppdOpenFile(PPDFile);
+
+  if (ppd)
+  {
+    ppd_size_t *size = ppd->sizes;
+
+    for (int k=0; k<ppd->num_sizes; k++)
+    {
+      dt_paper_info_t *paper = (dt_paper_info_t*)malloc(sizeof(dt_paper_info_t));
```
Comment (houz) [inline src/common/cups_print.c]: Move that into the if() to not leak it when the paper already exists.

File: src/common/cups_print.c [hunk 12]
```diff
@@ -0,0 +1,538 @@
...
+  }
+
+  if (pinfo->page.landscape)
+    num_options = cupsAddOption("landscape", "true", num_options, &options);
+  else
+    num_options = cupsAddOption("landscape", "false", num_options, &options);
+
+  // print lp options
+
+  if (darktable.unmuted & DT_DEBUG_PRINT)
```
Comment (houz) [inline src/common/cups_print.c]: Better use dt_print(DT_DEBUG_PRINT, ...)

File: src/common/pdf.h [hunk 13]
```diff
@@ -31,7 +31,9 @@
 #define dt_pdf_inch_to_point(inch)      ((inch) * 72.0)
 #define dt_pdf_point_to_inch(pt)        ((pt) / 72.0)
 #define dt_pdf_mm_to_point(mm)          dt_pdf_inch_to_point((mm) / 25.4)
+#define dt_pdf_point_to_mm(pt)          dt_pdf_point_to_inch(pt * 25.4)
```
Comment (houz) [inline src/common/pdf.h]: Better write `(pt)`

File: src/common/printprof.c [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+{
+  cmsHPROFILE hInProfile, hOutProfile;
+  cmsHTRANSFORM hTransform;
+  cmsUInt32Number wInput, wOutput;
+  int OutputColorSpace;
+
+  if(!g_file_test(profile, G_FILE_TEST_IS_REGULAR))
+    return 1;
+
+  void *out = (void *)malloc(width*height*3);
```
Comment (houz) [inline src/common/printprof.c]: Shouldn't there be bpp be used somehow? Or do you intend to print 8bpp only?

File: src/control/jobs/control_jobs.h [hunk 15]
```diff
@@ -32,6 +36,22 @@ typedef struct dt_control_export_t
   gboolean style_append;
 } dt_control_export_t;
 
+#ifdef HAVE_PRINT
+// note that dt_control_print_t must start as dt_control_export_t as it is also used by the export circuitry (inheritance)
```
Comment (houz) [inline src/control/jobs/control_jobs.h]: Just put a dt_control_export_t in the beginning then to be future proof.

File: src/imageio/storage/disk.c [hunk 16]
```diff
@@ -312,13 +312,27 @@ void init(dt_imageio_module_storage_t *self)
 
 void *get_params(dt_imageio_module_storage_t *self)
 {
+  const dt_view_t *cv = dt_view_manager_get_current_view(darktable.view_manager);
   dt_imageio_disk_t *d = (dt_imageio_disk_t *)calloc(1, sizeof(dt_imageio_disk_t));
 
-  char *text = dt_conf_get_string("plugins/imageio/storage/disk/file_directory");
-  g_strlcpy(d->filename, text, sizeof(d->filename));
-  g_free(text);
+  // we reuse the disk module for the print module, in the print view get the
```
Comment (houz) [inline src/imageio/storage/disk.c]: That sounds like a terrible idea.

File: src/libs/print_settings.c [hunk 17]
```diff
@@ -0,0 +1,1398 @@
...
+#include "common/variables.h"
+#include "common/cups_print.h"
+#include "common/image_cache.h"
+#include "common/pdf.h"
+#include "common/tags.h"
+#include "common/printprof.h"
+#include "dtgtk/resetlabel.h"
+#include "libs/lib.h"
+#include "gui/gtk.h"
+#include "gtk/gtk.h"
```
Comment (houz) [inline src/libs/print_settings.c]: that's `#include <gtk/gtk.h>` I guess? But since it gets pulled in by gui/gtk.h you can just omit it.

File: src/libs/print_settings.c [hunk 18]
```diff
@@ -0,0 +1,1398 @@
...
+    margin_h += ps->prt.printer.hw_margin_top + ps->prt.printer.hw_margin_bottom;
+  }
+
+  const int32_t width_pix = (width * ps->prt.printer.resolution) / 25.4;
+  const int32_t height_pix = (height * ps->prt.printer.resolution) / 25.4;
+
+  const double pa_width  = (width  - margin_w) / 25.4;
+  const double pa_height = (height - margin_h) / 25.4;
+
+  if (darktable.unmuted & DT_DEBUG_PRINT)
```
Comment (houz) [inline src/libs/print_settings.c]: dt_print() again

File: src/libs/print_settings.c [hunk 19]
```diff
@@ -0,0 +1,1398 @@
...
+
+  int max_width  = (pa_width  * ps->prt.printer.resolution);
+  int max_height = (pa_height * ps->prt.printer.resolution);
+
+  // make sure we are a multiple of four
+
+  max_width -= max_width % 4;
+  max_height -= max_height % 4;
+
+  const gchar *printer_profile = dt_conf_get_string("plugins/print/printer/iccprofile");
```
Comment (houz) [inline src/libs/print_settings.c]: I would prefer if those would be handed over in self. I know that we go throught he conf in many places, but that is not really a good design.

File: src/libs/print_settings.c [hunk 20]
```diff
@@ -0,0 +1,1398 @@
...
+  if (*printer_profile)
+    if (dt_apply_printer_profile(imgid, (void **)&(dat.ps->buf), dat.width, dat.height, dat.bpp, printer_profile, pintent))
+    {
+      dt_control_log(_("cannot apply printer profile `%s'"), printer_profile);
+      dt_control_queue_redraw();
+      return;
+    }
+
+  float border;
+  dt_pdf_parse_length("0 mm", &border);
```
Comment (houz) [inline src/libs/print_settings.c]: That's the same as setting border to 0. :)

File: src/libs/print_settings.c [hunk 21]
```diff
@@ -0,0 +1,1398 @@
...
+
+  const float page_width  = dt_pdf_mm_to_point(width);
+  const float page_height = dt_pdf_mm_to_point(height);
+
+  char filename[PATH_MAX] = { 0 };
+  snprintf(filename, sizeof(filename), "/tmp/pf_%d.pdf", imgid);
```
Comment (houz) [inline src/libs/print_settings.c]: Better use tempnam() to not get in trouble when printing the same image twice.

File: src/libs/print_settings.c [hunk 22]
```diff
@@ -0,0 +1,1398 @@
...
+*/
+  dt_pdf_image_t *pdf_image = dt_pdf_add_image(pdf, (uint8_t *)dat.ps->buf, dat.width, dat.height, 8, icc_id, border);
+
+  //  PDF bounding-box has origin on bottom-left
+  pdf_image->bb_x      = dt_pdf_pixel_to_point((float)margin_left, ps->prt.printer.resolution);
+  pdf_image->bb_y      = dt_pdf_pixel_to_point((float)margin_bottom, ps->prt.printer.resolution);
+  pdf_image->bb_width  = dt_pdf_pixel_to_point((float)iwidth, ps->prt.printer.resolution);
+  pdf_image->bb_height = dt_pdf_pixel_to_point((float)iheight, ps->prt.printer.resolution);
+
+  if (ps->prt.page.landscape && (dat.width > dat.height))
```
Comment (houz) [inline src/libs/print_settings.c]: What if you use portrait orientation and a landscape image?

File: src/libs/print_settings.c [hunk 23]
```diff
@@ -0,0 +1,1398 @@
...
+  else
+    pdf_image->rotate_to_fit = FALSE;
+
+  dt_pdf_page_t *pdf_page = dt_pdf_add_page(pdf, &pdf_image, 1);
+  dt_pdf_finish(pdf, &pdf_page, 1);
+
+  // free memory
+
+  free (dat.ps->buf);
+
```
Comment (houz) [inline src/libs/print_settings.c]: You leak printer_profile, pdf_image and pdf_page.

File: src/libs/print_settings.c [hunk 24]
```diff
@@ -0,0 +1,1398 @@
...
+static void _set_printer(dt_lib_module_t *self, const char *printer_name)
+{
+  dt_lib_print_settings_t *ps = (dt_lib_print_settings_t *)self->data;
+  dt_printer_info_t *printer = dt_get_printer_info (printer_name);
+
+  if (!printer) return;
+
+  memcpy(&ps->prt.printer, printer, sizeof(dt_printer_info_t));
+
+  // if there is 0 hardware margins, set de user marging to 20mm
```
Comment (houz) [inline src/libs/print_settings.c]: You say 20 here but use 15 below.

File: src/libs/print_settings.c [hunk 25]
```diff
@@ -0,0 +1,1398 @@
...
+  }
+  if (ps->prt.printer.hw_margin_right == 0)
+  {
+    ps->prt.page.margin_right = 15;
+    gtk_spin_button_set_value(GTK_SPIN_BUTTON(ps->b_right), ps->prt.page.margin_right);
+  }
+
+  dt_conf_set_string("plugins/print/print/printer", printer_name);
+
+  const char *default_paper = dt_conf_get_string("plugins/print/print/paper");
```
Comment (houz) [inline src/libs/print_settings.c]: it's not const and you are leaking it.

File: src/libs/print_settings.c [hunk 26]
```diff
@@ -0,0 +1,1398 @@
...
+
+  GList *printers = dt_get_printers();
+  int np=0;
+  int printer_index = 0;
+  char printer_name[128] = { 0 };
+
+  label = dt_ui_section_label_new(_("printer"));
+  gtk_box_pack_start(GTK_BOX(self->widget), label, TRUE, TRUE, 0);
+
+  comb = GTK_COMBO_BOX(gtk_combo_box_text_new());
```
Comment (houz) [inline src/libs/print_settings.c]: Better use a bauhaus combobox

File: src/libs/print_settings.c [hunk 27]
```diff
@@ -0,0 +1,1398 @@
...
+  //  Add printer profile combo
+
+  d->pprofile = dt_bauhaus_combobox_new(NULL);
+  dt_bauhaus_widget_set_label(d->pprofile, NULL, _("profile"));
+
+  int combo_idx = -1, n=0;
+  GList *l = d->profiles;
+
+  gtk_box_pack_start(GTK_BOX(self->widget), GTK_WIDGET(d->pprofile), TRUE, TRUE, 0);
+  const gchar *printer_profile = dt_conf_get_string("plugins/print/printer/iccprofile");
```
Comment (houz) [inline src/libs/print_settings.c]: printer_profile is leaking.

File: src/libs/print_settings.c [hunk 28]
```diff
@@ -0,0 +1,1398 @@
...
+  gtk_box_pack_start(GTK_BOX(self->widget), GTK_WIDGET(d->papers), TRUE, TRUE, 0);
+
+  // now set recorded default/active printer
+
+  gtk_combo_box_set_active(GTK_COMBO_BOX(comb), printer_index);
+  _set_printer(self, printer_name);
+
+  //// portrait / landscape
+
+  d->landscape = gtk_radio_button_new_with_label(NULL, _("landscape"));
```
Comment (houz) [inline src/libs/print_settings.c]: Use a bauhaus combobox instead, label "orientation" and the two options "landscape" and "portrait".

File: src/libs/print_settings.c [hunk 29]
```diff
@@ -0,0 +1,1398 @@
...
+  d->portrait = gtk_radio_button_new_with_label_from_widget(GTK_RADIO_BUTTON (d->landscape), _("portrait"));
+  GtkBox *hbox2 = GTK_BOX(gtk_box_new(GTK_ORIENTATION_HORIZONTAL, DT_PIXEL_APPLY_DPI(5)));
+  gtk_box_pack_start(GTK_BOX(hbox2), GTK_WIDGET(d->landscape), TRUE, TRUE, 0);
+  gtk_box_pack_start(GTK_BOX(hbox2), GTK_WIDGET(d->portrait), TRUE, TRUE, 0);
+
+  comb = GTK_COMBO_BOX(gtk_combo_box_text_new());
```
Comment (houz) [inline src/libs/print_settings.c]: bauhaus again

File: src/libs/print_settings.c [hunk 30]
```diff
@@ -0,0 +1,1398 @@
...
+
+  //  Add export profile combo
+
+  d->profile = dt_bauhaus_combobox_new(NULL);
+  dt_bauhaus_widget_set_label(d->profile, NULL, _("profile"));
+
+  gtk_box_pack_start(GTK_BOX(self->widget), GTK_WIDGET(d->profile), TRUE, TRUE, 0);
+  dt_bauhaus_combobox_add(d->profile, _("image settings"));
+
+  const gchar *iccprofile = dt_conf_get_string("plugins/print/print/iccprofile");
```
Comment (houz) [inline src/libs/print_settings.c]: Leak

File: src/libs/print_settings.c [hunk 31]
```diff
@@ -0,0 +1,1398 @@
...
+
+  //  Add export style combo
+
+  d->style = dt_bauhaus_combobox_new(NULL);
+  dt_bauhaus_widget_set_label(d->style, NULL, _("style"));
+
+  dt_bauhaus_combobox_add(d->style, _("none"));
+
+  GList *styles = dt_styles_get_list("");
+  const gchar *current_style = dt_conf_get_string("plugins/print/print/style");
```
Comment (houz) [inline src/libs/print_settings.c]: Leak

File: src/common/cups_print.c (hunk 6 above)
Comment (TurboGit) [inline src/common/cups_print.c]: Indeed, it is reporting the size in 1/72 inch.

File: src/common/cups_print.c (hunk 7 above)
Comment (TurboGit) [inline src/common/cups_print.c]: Right, we will have to move to IPP in the future. There is not enough doc or code sample to move to IPP now, some parts are really not clear. This has been deprecated very recently.

File: src/common/cups_print.c (hunk 8 above)
Comment (TurboGit) [inline src/common/cups_print.c]: Indeed a copy is passed, actually a symlink in /tmp to the corresponding PPD.

File: src/common/cups_print.c (hunk 9 above)
Comment (TurboGit) [inline src/common/cups_print.c]: Ok, fixed.

File: src/common/cups_print.c (hunk 10 above)
Comment (TurboGit) [inline src/common/cups_print.c]: Indeed, fixed.

File: src/common/cups_print.c (hunk 11 above)
Comment (TurboGit) [inline src/common/cups_print.c]: Good catch!

File: src/common/pdf.h (hunk 13 above)
Comment (TurboGit) [inline src/common/pdf.h]: Fixed.

File: src/common/printprof.c (hunk 14 above)
Comment (TurboGit) [inline src/common/printprof.c]: Only 8 bpp as this is what the CUPS print driver can actually handle. So in input we have either 8bit or 16bit but in all cases the output is 8bit.

File: src/imageio/storage/disk.c (hunk 16 above)
Comment (TurboGit) [inline src/imageio/storage/disk.c]: Right :) All this has been removed in version based on PDF.

File: src/control/jobs/control_jobs.h (hunk 15 above)
Comment (TurboGit) [inline src/control/jobs/control_jobs.h]: Hum, all this is left over from old implementation. Will remove an squash it!

File: src/libs/print_settings.c (hunk 17 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: Removed.

File: src/libs/print_settings.c (hunk 20 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: Indeed, I suppose that's some left-over of testing.

File: src/libs/print_settings.c (hunk 22 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: That's fine, the image is always on the same orientation on screen. I was puzzled when I implemented that but it is correct.

File: src/libs/print_settings.c (hunk 23 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: Fixed.

File: src/libs/print_settings.c (hunk 24 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: Comment fixed;

File: src/libs/print_settings.c (hunk 25 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: Indeed, fixed.

File: src/libs/print_settings.c (hunk 27 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: Fixed.

File: src/views/print.c [hunk 32]
```diff
@@ -0,0 +1,321 @@
...
+  cairo_rectangle (cr, px, py, pwidth, pheight);
+  cairo_fill (cr);
+
+  // display non-printable area
+  cairo_set_source_rgb (cr, 0.1, 0.1, 0.1);
+
+  // top-left
+  int npx = px + (np_left / pg_width) * pwidth;
+  int npy = py + (np_top / pg_height) * pheight;
+  cairo_move_to (cr, npx, npy); cairo_line_to (cr, npx-10, npy);
```
Comment (houz) [inline src/views/print.c]: That could be a single cairo_move_to() and two cairo_line_to().

File: src/libs/print_settings.c (hunk 30 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: Fixed.

File: src/libs/print_settings.c (hunk 31 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: Fixed.

File: src/common/cups_print.c (hunk 12 above)
Comment (TurboGit) [inline src/common/cups_print.c]: Not sure to understand what you mean here?

File: src/libs/print_settings.c [hunk 33]
```diff
@@ -0,0 +1,1398 @@
...
+      strncpy(printer_name,printer->name,sizeof(printer_name));
+    }
+    printers = g_list_next (printers);
+    np++;
+  }
+
+  gtk_box_pack_start(GTK_BOX(self->widget), GTK_WIDGET(comb), TRUE, TRUE, 0);
+  d->printers = comb;
+  g_signal_connect(G_OBJECT(d->printers), "changed", G_CALLBACK(_printer_changed), self);
+  g_list_free (printers);
```
Comment (houz) [inline src/libs/print_settings.c]: That should be `g_list_free_full(printers, free);`

File: src/libs/print_settings.c [hunk 34]
```diff
@@ -0,0 +1,1398 @@
...
+
+  // next add corresponding papers
+
+  // first clear current list
+
+  dt_bauhaus_combobox_clear(ps->papers);
+
+  // then add papers for the given printer
+
+  ps->paper_list = dt_get_papers (printer_name);
```
Comment (houz) [inline src/libs/print_settings.c]: you never free that (using g_list_free_full(..., free);)

File: src/libs/print_settings.c (hunk 28 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: Done.

File: src/libs/print_settings.c (hunk 29 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: Done.

File: src/libs/print_settings.c (hunk 33 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: Fixed.

File: src/libs/print_settings.c (hunk 34 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: Fixed.

File: src/libs/print_settings.c (hunk 21 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: Indeed, fixed.

File: src/libs/print_settings.c (hunk 26 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: Done.

File: src/common/cups_print.c (hunk 12 above)
Comment (houz) [inline src/common/cups_print.c]: To not use normal printf, guarded by an if, but just using dt_print().

File: src/common/printprof.c [hunk 35]
```diff
@@ -0,0 +1,92 @@
...
+      cmsDoTransform(hTransform, (const void *)&ptr_in[k*width*3], (void *)&ptr_out[k*width*3], width);
+  }
+
+  cmsDeleteTransform(hTransform);
+
+  free(*in);
+  *in = out;
+
+  return 0;
+}
```
Comment (houz) [inline src/common/printprof.c]: this file misses the modelines.

File: src/develop/lightroom.c [hunk 36]
```diff
@@ -192,18 +192,18 @@ typedef struct dt_iop_bilat_params_t
 #define LRDT_COLORIN_VERSION 1
 #define DT_IOP_COLOR_ICC_LEN 100
 
-typedef enum dt_iop_color_intent_t
+typedef enum dtlr_iop_color_intent_t
```
Comment (houz) [inline src/develop/lightroom.c]: Why did you rename them? It would make more sense to move the intent definition from the two places (librhtoom.c + color.h) to a single place like common/colorspaces.h and use it everywhere.

File: src/common/cups_print.c [hunk 37]
```diff
@@ -0,0 +1,538 @@
...
+  else
+    dt_control_log(_("printing image %d on `%s'"), imgid, pinfo->printer.name);
+
+  cupsFreeOptions (num_options, options);
+}
+
+static void _get_image_dimension (int32_t imgid, int32_t *iwidth, int32_t *iheight)
+{
+  dt_mipmap_buffer_t buf;
+  dt_mipmap_cache_get(darktable.mipmap_cache, &buf, imgid, DT_MIPMAP_3, DT_MIPMAP_BEST_EFFORT, 'r');
```
Comment (houz) [inline src/common/cups_print.c]: This is not going to work, mipmap 3 has a fixed size that you can't get the image size from.

Comment (TurboGit) [inline src/common/cups_print.c]: Right but here we just need the aspect more than the actual size. To display on-screen we just need to computer the origin.

File: src/common/printprof.c (hunk 35 above)
Comment (TurboGit) [inline src/common/printprof.c]: fixed.

File: src/develop/lightroom.c (hunk 36 above)
Comment (TurboGit) [inline src/develop/lightroom.c]: Indeed, lot better!

File: src/common/cups_print.c (hunk 37 above)
Comment (LebedevRI) [inline src/common/cups_print.c]: mipmap 3 has a fixed size, so you will *always* get the same aspect ratio, and it is not guaranteed to be be equal to the actual aspect ratio of the image.

File: src/common/cups_print.c (hunk 12 above)
Comment (TurboGit) [inline src/common/cups_print.c]: I see, didn't know that dt_print() exists! Fixed.

File: src/common/cups_print.c (hunk 37 above)
Comment (TurboGit) [inline src/common/cups_print.c]: This is not what I'm seeing. This is used to get the proper displayed image on the print module and it works. I have just added a printf for buf.width/buf.height and I get different values.

Comment (TurboGit) [inline src/common/cups_print.c]: Is there any other way to get the image aspect ratio?

File: src/libs/print_settings.c (hunk 19 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: I've changed that too.

File: src/common/cups_print.c (hunk 37 above)
Comment (LebedevRI) [inline src/common/cups_print.c]: dt_dev_pixelpipe_get_dimensions()

Comment (TurboGit) [inline src/common/cups_print.c]: Great! Working like a charm! Thanks.

File: src/common/printprof.c (hunk 3 above)
Comment (TurboGit) [inline src/common/printprof.c]: left over of previous implementation. Now fixed.

File: src/views/view.c (hunk 5 above)
Comment (TurboGit) [inline src/views/view.c]: Ok! Thanks.

File: src/views/print.c (hunk 32 above)
Comment (TurboGit) [inline src/views/print.c]: Indeed, fixed. I missed that in my first batch of fixes.

File: src/libs/print_settings.c [hunk 38]
```diff
@@ -0,0 +1,1476 @@
...
+           margin_top, margin_bottom, margin_left, margin_right);
+
+  // we have the exported buffer, let's apply the printer profile
+
+  if (*ps->v_piccprofile)
+    if (dt_apply_printer_profile(imgid, (void **)&(dat.ps->buf), dat.width, dat.height, dat.bpp, ps->v_piccprofile, ps->v_pintent))
+    {
+      dt_control_log(_("cannot apply printer profile `%s'"), ps->v_piccprofile);
+      dt_control_queue_redraw();
+      return;
```
Comment (houz) [inline src/libs/print_settings.c]: returning here will leak ps->buf which got allocated in write_image().

File: src/libs/print_settings.c [hunk 39]
```diff
@@ -0,0 +1,1476 @@
...
+    dt_conf_set_string("plugins/print/print/style", "");
+  }
+  else
+  {
+    const gchar *style = dt_bauhaus_combobox_get_text(ps->style);
+    dt_conf_set_string("plugins/print/print/style", style);
+  }
+
+  if (ps->v_style) g_free(ps->v_style);
+  ps->v_style = g_strdup(dt_conf_get_string("plugins/print/print/style"));
```
Comment (houz) [inline src/libs/print_settings.c]: the g_strdup should be removed, dt_conf_get_string() already returns a newly allocated string.

File: src/libs/print_settings.c [hunk 40]
```diff
@@ -0,0 +1,1476 @@
...
+  GList *prof = ps->profiles;
+  while(prof)
+  {
+    // could use g_list_nth. this seems safer?
+    dt_lib_export_profile_t *pp = (dt_lib_export_profile_t *)prof->data;
+    if(strcmp(pp->name,printer_profile)==0)
+    {
+      dt_conf_set_string("plugins/print/printer/iccprofile", pp->filename);
+      if (ps->v_piccprofile) g_free(ps->v_piccprofile);
+      ps->v_piccprofile = pp->filename;
```
Comment (houz) [inline src/libs/print_settings.c]: shouldn't that be g_strdup'ed?

File: src/libs/print_settings.c [hunk 41]
```diff
@@ -0,0 +1,1476 @@
...
+  int np=0;
+  int printer_index = 0;
+  char printer_name[128] = { 0 };
+
+  label = dt_ui_section_label_new(_("printer"));
+  gtk_box_pack_start(GTK_BOX(self->widget), label, TRUE, TRUE, 0);
+
+  d->printers = dt_bauhaus_combobox_new(NULL);
+
+  const char *default_printer = dt_conf_get_string("plugins/print/print/printer");
```
Comment (houz) [inline src/libs/print_settings.c]: default_printer is leaking

File: src/views/view.c (hunk 4 above)
Comment (houz) [inline src/views/view.c]: slideshow is not optional, it will always be built.

Comment (LebedevRI) [inline src/views/view.c]: Right, i have seen
option(BUILD_SLIDESHOW "Build the opengl slideshow viewer" ON)
(https://github.com/darktable-org/darktable/blob/5d7890b9cdf01a92a5755749e1537eec3a4d6229/CMakeLists.txt#L17)
and thought that was about slideshow view, and not about that old dtview thingy

File: src/libs/print_settings.c (hunk 39 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: Fixed.

File: src/libs/print_settings.c (hunk 41 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: Fixed.

File: src/libs/print_settings.c (hunk 38 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: Indeed. Fixed too. I'll do a final pass and will merge.

File: src/libs/print_settings.c (hunk 40 above)
Comment (TurboGit) [inline src/libs/print_settings.c]: Yep, this call was missing indeed!

**Comment by houz** [comment]
Great, I think we are in a state where you can bring this to a mergeable state so we can bring this to wider testing.

**Comment by TurboGit** [comment]
Thanks for all the review! Will merge now. I'm now ready for bug fixes from the field :)

**Comment by houz** [comment]
I have no idea what you did to the git history, but to me it looks really strange. I can hardly do any bisecting on the resulting tree and your commit dates look off, too. How did you merge the branch into master?

**Comment by TurboGit** [comment]
I have done a standard:

$ git merge --log print-module

File: src/common/printprof.c [hunk 42]
```diff
@@ -0,0 +1,96 @@
...
+
+  cmsCloseProfile(hInProfile);
+  cmsCloseProfile(hOutProfile);
+
+  if (bpp == 8)
+  {
+    const uint8_t *ptr_in = (uint8_t *)*in;
+    uint8_t *ptr_out = (uint8_t *)out;
+
+    for (int k=0; k<height; k++)
```
Comment (LebedevRI) [inline src/common/printprof.c]: (I noticed this only now.)
Here, and in *else* branch, something like this (list in shared() might need correctio)
```
#ifdef _OPENMP
#pragma omp parallel for schedule(static) default(none) shared(ptr_in, ptr_out, hTransform, height, k, width)
#endif
```
could be added before *for*.
That would result in faster color transform, since LCMS, internally, does not do any parallelization.

Comment (TurboGit) [inline src/common/printprof.c]: I'll check. k is not needed here as shared() never include the loop variable.
